src/main/java/io/vertx/core/impl/ContextBase.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.core.impl;
+
+import io.vertx.core.spi.context.ContextKey;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.function.Supplier;
+
+/**
+ * Base class for context.
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+class ContextBase {
+
+  private static final VarHandle LOCALS_UPDATER = MethodHandles.arrayElementVarHandle(Object[].class);
+
+  final Object[] locals;
+
+  ContextBase(Object[] locals) {
+    this.locals = locals;
+  }
+
+  public final <T> T getLocal(ContextKey<T> key) {
+    ContextKeyImpl<T> internalKey = (ContextKeyImpl<T>) key;
+    int index = internalKey.index;
+    if (index >= locals.length) {
+      throw new IllegalArgumentException();
+    }
+    Object res = LOCALS_UPDATER.getVolatile(locals, index);
+    return (T) res;
+  }
+
+  public final <T> T getLocal(ContextKey<T> key, Supplier<T> supplier) {
+    ContextKeyImpl<T> internalKey = (ContextKeyImpl<T>) key;
+    int index = internalKey.index;
+    if (index >= locals.length) {
+      throw new IllegalArgumentException();
+    }
+    Object res;
+    while (true) {
+      res = LOCALS_UPDATER.getVolatile(locals, index);
+      if (res != null) {
+        break;
+      }
+      Object initial = supplier.get();
+      if (initial == null) {
+        throw new IllegalStateException();
+      }
+      if (LOCALS_UPDATER.compareAndSet(locals, index, null, initial)) {
```

Comment:
We don't need atomicity here, but the same semantic of putLocal, which doesn't care of races, imo.
If we need that level of atomicity we can have a new method which ensure an atomic change and use the compare and set.

Reply:
If I understand correctly, the goal of this method is to initialize the value if not present.

In this case:

- we can't do it with the same semantics as `putLocal`, can we?
- how about changing `compareAndSet` with `compareAndExchange` and return the value directly? This would avoid having to invoke `getVolatile` again

Reply:
I think I just applied a common pattern here that initialises the state when it is null otherwise discard the value obtained from the supplier if it was set by another thread concurrently, e.g in reactiverse contextual logging, this will be initialised with a concurrent hash map:

```java
Map<String, Object> map = getLocal(key, () -> new ConcurrentHashMap<>());
```

It is similar to `Map#computeIfAbsent(K, Function<K, V>)`

Reply:
@tsegismont compareAndExchange only exist since Java 9, I do intent to back port this to 4.x branch for deprecation purpose (and actually the existing concurrent map would be retrofit as a context key)